### PR TITLE
Motion controller: Roboclaw Support

### PR DESCRIPTION
**Pull request by dringakn**
@khancyr this is a working version, I have tested it today on a real roboclaw.

### CODE REVIEW

File: libraries/SRV_Channel/SRV_Channels.cpp [hunk 1]
```diff
@@ -134,10 +134,14 @@ const AP_Param::GroupInfo SRV_Channels::var_info[] = {
...
 
+    // @Group: _MOCR_
+    // @Path: ../AP_MotionController/AP_MotionController.cpp
+    AP_SUBGROUPINFO(volz, "_MOCR_",  21, SRV_Channels, AP_MotionController),
+
 #if HAL_SUPPORT_RCOUT_SERIAL
     // @Group: _BLH_
     // @Path: ../AP_BLHeli/AP_BLHeli.cpp
-    AP_SUBGROUPINFO(blheli, "_BLH_",  21, SRV_Channels, AP_BLHeli),
+    AP_SUBGROUPINFO(blheli, "_BLH_",  22, SRV_Channels, AP_BLHeli),
```
Comment (rmackay9) [inline libraries/SRV_Channel/SRV_Channels.cpp]: Oh, can't do this!  We can't move the position of existing parameters or they become corrupted.  So the MOCR params should be moved to a free spot normally at the end of the list.

File: APMrover2/system.cpp [hunk 2]
```diff
@@ -145,7 +145,8 @@ void Rover::init_ardupilot()
     startup_ground();
 
     Mode *initial_mode = mode_from_mode_num((enum Mode::Number)g.initial_mode.get());
-    if (initial_mode == nullptr) {
```
Comment (rmackay9) [inline APMrover2/system.cpp]: normally we ask that formatting changes be put into a separate commit.  This one to move the bracket to the next line doesn't actually fit with our coding convention (there's a page on the developer wiki) so we should back out this change.

File: libraries/AP_MotionController/AP_MotionController.cpp [hunk 3]
```diff
@@ -15,22 +15,22 @@
...
-#include <AP_MotionController/AP_MotionController.h>
+#include <GCS_MAVLink/GCS.h>
+#include "AP_MotionController.h"
 
 extern const AP_HAL::HAL &hal; // External reference for console debugging
 
 const AP_Param::GroupInfo AP_MotionController::var_info[] = {
     // @Param: MASK
-    // @DisplayName: MOCR_MASK
+    // @DisplayName: SERVO_MOCR_MASK
```
Comment (rmackay9) [inline libraries/AP_MotionController/AP_MotionController.cpp]: The Display name shouldn't be the same as the parameter name.  I'd say it should probably be something like "RoboClaw motion controller board enable mask" or something to that effect.

File: libraries/AP_MotionController/AP_MotionController.cpp [hunk 4]
```diff
@@ -45,13 +45,36 @@ AP_MotionController::~AP_MotionController()
...
     //gcs().send_text(MAV_SEVERITY_INFO, "AP_MotionController Initialized.");
 }
 
-void AP_MotionController::update(void)
+double AP_MotionController::constrain_map(double x, double in_min, double in_max, double out_min, double out_max)
 {
-    //hal.console->printf("AP_MotionController updated.\n");
-    //gcs().send_text(MAV_SEVERITY_INFO, "AP_MotionController Updated.");
+    //constrain the input value between in_min and in_max
+    if (x < in_min)
```
Comment (rmackay9) [inline libraries/AP_MotionController/AP_MotionController.cpp]: minor formatting thing but normally we have a bracket even for cases where there's just a single line statement afterwards.

File: libraries/AP_MotionController/AP_MotionController.cpp [hunk 5]
```diff
@@ -45,13 +45,36 @@ AP_MotionController::~AP_MotionController()
...
+    static int counter = 0;
+    uint32_t m1 = AP_MotionController::constrain_map(steering, SRV_Channels::srv_channel(1)->get_output_min(), SRV_Channels::srv_channel(1)->get_output_max(), 0, 127);
+    uint32_t m2 = AP_MotionController::constrain_map(steering, SRV_Channels::srv_channel(3)->get_output_min(), SRV_Channels::srv_channel(3)->get_output_max(), 0, 127);
+    roboclaw.ForwardBackwardM1(AP_RoboClaw::ROBOCLAW_ADDR, m1);
+    roboclaw.ForwardBackwardM2(AP_RoboClaw::ROBOCLAW_ADDR, m2);
+    //roboclaw.SpeedM1M2(AP_RoboClaw::ROBOCLAW_ADDR, steering, throttle);
+    if (++counter >= 10) // Call every fifth of a second (50Hz calling rate)
+    {
+        hal.console->printf("Time:%05.3f Steering:%d Throttle:%d\n", AP_HAL::millis64() / 1000.0, m1, m2);
+        //port->printf("Time:%05.3f Steering:%d Throttle:%d\n", AP_HAL::millis64()/1000.0, steering, throttle);
```
Comment (rmackay9) [inline libraries/AP_MotionController/AP_MotionController.cpp]: It's best to remove commented out code.  It just becomes messy if we leave this in.  For debug, you'll find examples where other libraries have a DEBUG_OUTPUT (or something like that) if you really want to leave the debug in there for future development.  Personally I'd say just remove it though.

**Comment by rmackay9** [comment]
Very interesting to have this.  I think the RoboClaw controllers are pretty popular so having support for their serial protocol is certainly interesting.  Can the RoboClaw controllers provide any feedback?

I think having this code in a new folder just below "library" is best though.  I think it should be somewhere else.. but I'm not sure where.  I also think if possible it should be enabled (in Rover) using the existing MOT_PWM_TYPE parameter.  Maybe @tridge knows where this library best belongs.

The name "AP_MotionController" should be made more specific I think.  It should probably be "RoboClaw_MotionController" or something similar.  I certainly think "RoboClaw" should be in the name.

**Comment by khancyr** [comment]
Hello thanks for the contribution!

I won't have much time to work on time on the next two weeks. But here is some point to improve. 

It should belong to a new class with a backend and a front-end since there are plenties of motor controler protocol. That will also allow us to get back feedback from the controler. And we should be able to select the control type (RC, PWM, velocity etc.). 

All the style change should avoid.

File: libraries/AP_MotionController/AP_MotionController.cpp [hunk 6]
```diff
@@ -0,0 +1,758 @@
...
+}
+
+AP_MotionController::~AP_MotionController()
+{
+}
+
+void AP_MotionController::init(void)
+{
+    AP_SerialManager &serial_manager = AP::serialmanager();
+    port = serial_manager.find_serial(AP_SerialManager::SerialProtocol_ESCTelemetry, 0);
```
Comment (OXINARF) [inline libraries/AP_MotionController/AP_MotionController.cpp]: Why ESCTelemetry?

File: libraries/AP_MotionController/AP_MotionController.cpp [hunk 7]
```diff
@@ -0,0 +1,758 @@
...
+    // Reset roboclaw communication port
+    if (bSuccess == false)
+    {
+        roboclaw.init(nullptr, 10000);
+    }
+
+    const double qppr = QPPR.get();
+    const double qpps_vmax = Vmax.get() * qppr / (2 * M_PI * Rw.get()); // Maximum linear velocity interm of qpps
+    const double qp_amax = Amax.get() * DEG_TO_RAD * qppr / (2 * M_PI); // Maximum angle interm of quadratur pulses (qp)
+    const double dz = DZ.get() * qp_amax / 100.0;                       // Dead zone around target position
```
Comment (OXINARF) [inline libraries/AP_MotionController/AP_MotionController.cpp]: Is double really necessary? We want to avoid it as much as possible, as the most used boards (STM32F4) don't have hardware support for it.

**Comment by davidbuzz** [comment]
no chances since review in 2018. appears abandoned.

**Comment by khancyr** [comment]
This need a full rework. Feels free to reopen a new PR. Thanks anyway
